smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+                    for (MemberShape memberShape : model.expectShape(operationShape.getInputShape()).members()) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    }
+                }
+
+                Shape outputPropertiesShape = getOutputPropertiesShape(operationShape);
+                operationToOutputPropertiesShape.put(operationShapeId, outputPropertiesShape.getId());
+                for (MemberShape memberShape : outputPropertiesShape.members()) {
+                    if (identifierIndex.getOperationOutputBindings(resourceShape, operationShape).values()
+
+                            .contains(memberShape.getMemberName())) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    } else {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(),
+                                doesNotRequireProperty(memberShape));
+                    }
+                    if (doesMemberShapeRequireProperty(memberShape.getId())
+                            || propertyNames.contains(memberShape.getMemberName())) {
+                        memberShapeToPropertyName.put(memberShape.getId(), getPropertyTraitName(memberShape)
+                            .orElse(memberShape.getMemberName()));
+                    }
+                }
+                // nesting is taking place, so index top level input/output members as not property.
+                if (!outputPropertiesShape.getId().equals(operationShape.getOutputShape())) {
+                    for (MemberShape memberShape : model.expectShape(operationShape.getOutputShape()).members()) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    }
+                }
+            }
+        }
+    }
+
+    public static PropertyBindingIndex of(Model model) {
+        return model.getKnowledge(PropertyBindingIndex.class, PropertyBindingIndex::new);
+    }
+
+    /**
+     * Gets the property name for a given member shape. Returns empty optional if the
+     * member shape does not correspond to a property.
+     *
+     * @param memberShapeId the ShapeId of the member shape to get the property name for.
+     * @return the property name for a given member shape if there is one.
+     */
+    public Optional<String> getPropertyName(ShapeId memberShapeId) {
+        return Optional.ofNullable(memberShapeToPropertyName.get(memberShapeId));
+    }
+
+    /**
+     * Returns true if a member shape positively maps to a property.
+     *
+     * {@see PropertyBindingIndex#getPropertyName(ShapeId)} will return a non-empty Optional if this method
+     * returns true.
+     *
+     * @param memberShapeId the ShapeId of the member shape to check
+     * @return true if member shape maps to a property on the given resource
+     */
+    public boolean isMemberShapeProperty(ShapeId memberShapeId) {
```

Comment:
I think this should ask for a MemberShape rather than ShapeId. That will give us flexibility to compute this stuff on demand in the future if we ever decide to (which we might after we profile this).

Also, doesn't this need a resource shape too, in the case a shape is reused?

Reply:
A resource shape is not needed as it's impossible for a single member shape to be a property in one resource, and not in another. Where it comes close to seeming valid is if one resource has an identifier that matches a required member and triggers an implicit identifier binding (trumping property binding). The reason this (should) fail is because identifiers are not allowed to match the name of properties.

It's worth a new test to confirm this is not allowed, but I cannot think of any case where a given member shape shared across two resources can, and should be allowed to map differently as resource properties.